### PR TITLE
8272093: Extract evacuation failure injection from G1CollectedHeap

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that extracts out the evacuation failure injection mechanism from `G1CollectedHeap` into its own class?
This reduces the amount of code in `G1CollectedHeap` a bit, moving somewhat uninteresting but somewhat important (testing) code out from there. Also documented it a bit. The other reason was that imho it stuck out a bit due to the ifdef's in `G1CollectedHeap`.

I checked that gcc does compile out the call to `evacuation_should_fail()` in `G1ParScanThreadState` in product mode.

Testing: tier1, local testing

Thanks,
  Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272093](https://bugs.openjdk.java.net/browse/JDK-8272093): Extract evacuation failure injection from G1CollectedHeap

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to 9464925781e0b03c84162b6568bbbb5a47572378
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 9464925781e0b03c84162b6568bbbb5a47572378

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5047/head:pull/5047` \
`$ git checkout pull/5047`

Update a local copy of the PR: \
`$ git checkout pull/5047` \
`$ git pull https://git.openjdk.java.net/jdk pull/5047/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5047`

View PR using the GUI difftool: \
`$ git pr show -t 5047`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5047.diff">https://git.openjdk.java.net/jdk/pull/5047.diff</a>

</details>

### CODE REVIEW

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1YoungGCEvacFailureInjector.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  // Should we inject evacuation failures in the current GC.
+  bool _inject_evacuation_failure_for_current_gc;
+
+  // Records the number of the last collection when evacuation failure happened.
+  // Used to determine whether evacuation failure injection should be in effect
+  // for the current GC.
+  size_t _last_collection_with_evacuation_failure;
+
+  // The number of evacuations between induced failures.
+  volatile size_t _evacuation_failure_object_count;
```
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1YoungGCEvacFailureInjector.hpp:49]: Any reason to not guard these with `#ifndef PRODUCT`?

**Review by kstefanj — APPROVED** [review]
Looks good, but I think the members should be left out in product builds.

File: src/hotspot/share/gc/g1/g1YoungGCEvacFailureInjector.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1YoungGCEvacFailureInjector.hpp:49]: Will do. I need to merge with master anyway.

**Comment by tschatzl** [comment]
Thanks @kstefanj @albertnetymk for your reviews.

/integrate

**Pull request closed without merging**